### PR TITLE
arch/cortex-m: Add support for `dsb` instruction

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

This pull request adds a `dsb()` function (Data Synchronization Barrier) function to the cortex-m support crate.

Pulled from #1220 

### Testing Strategy

This pull request was tested by...

### TODO or Help Wanted

This pull request still needs...

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make formatall`.

### CODE REVIEW

File: arch/cortex-m/src/support.rs [hunk 1]
```diff
@@ -24,6 +24,17 @@ pub unsafe fn wfi() {
 /// WFI instruction (mock)
 pub unsafe fn wfi() {}
 
+#[cfg(target_os = "none")]
+#[inline(always)]
+/// DSB instruction
+pub unsafe fn dsb() {
+    asm!("dsb 0xf" :::: "volatile");
```
Comment (brghena) [inline arch/cortex-m/src/support.rs]: What does the `0xf` do here? As far as I can tell, `dsb` accepts no arguments.

**Comment by rajivr** [comment]
[AN321 (ARM Cortex-M Programming Guide to
Memory Barrier Instructions)](https://developer.arm.com/docs/dai0321/latest) and Table 5.42/5.43 of Yiu describes scenarios where we would need memory barrier instructions. In STM32 port, I am using `dsb` as recommended in section 4.14 (Entering sleep modes) of AN321.

**Comment by ppannuto** [comment]
The note below that table says:

In simple designs that do no include a system level write buffer, the
memory barrier instruction
is not required before entering sleep mode in the Cortex-M3 (r2p0 or
later), Cortex-M4,
Cortex-M0+, and Cortex-M0 processors. This is handled by the processor.

Do we actually need to add DSB in instructions in that case?

**Comment by alevy** [comment]
@ppannuto are you basically asking whether the STM32 has a write-buffer? (I don't know the answer, btw)

**Comment by rajivr** [comment]
@ppannuto: Your are right. I just re-checked the STM32F4XX vendor HAL code. It looks like we do not need `dsb` for STM32F4XX when entering sleep. I'll update #1220 accordingly.

**Pull request closed without merging**